bitcoin/src/pow.rs

```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+/// Error from `get_block_header_by_height` callback.
+///
+/// Occurse when there is an error while get a new block header
+/// from an external source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
+
+impl fmt::Display for GetBlockHeaderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            Self::Generic(error) => error,
+            Self::NotFound => "Block header not found",
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for GetBlockHeaderError {}
+
+/// Calculate the next Work required as the `GetNextWorkRequired` do
+/// in bitcoin core.
+///
+/// Return a error if the callback provide from the caller fails.
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensus_params: &Params,
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    mut get_block_header_by_height: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: FnMut(u32) -> Result<Header, GetBlockHeaderError>,
+{
+    let adjustment_interval = consensus_params.difficulty_adjustment_interval();
+    let pow_target_timespan = consensus_params.pow_target_timespan;
+    let no_pow_retargeting = consensus_params.no_pow_retargeting;
+    let allow_min_difficulty_blocks = consensus_params.allow_min_difficulty_blocks;
+    let pow_target_spacing = consensus_params.pow_target_spacing as u32;
+
+    let current_height = current_height.to_consensus_u32() as u64;
+    let current_time = current_header.time;
+
+    let last_adjustment_height = current_height - (adjustment_interval - 1);
+    let last_adjustment_block = get_block_header_by_height(last_adjustment_height as u32)?;
+    let last_adjustment_time = last_adjustment_block.time;
+
+    if (current_height + 1) % adjustment_interval != 0 {
+        if allow_min_difficulty_blocks {
+            let pow_limit = consensus_params.pow_limit.to_compact_lossy().to_consensus();
+            if new_block_timestamp > current_header.time + pow_target_spacing * 2 {
+                return Ok(pow_limit);
+            } else {
+                use hashes::Hash;
+                let mut header = current_header;
+                let mut height = current_height;
+                while header.prev_blockhash != BlockHash::all_zeros()
+                    && height % adjustment_interval != 0
+                    && header.bits.to_consensus() == pow_limit
```

Comment:
You're converting everything to consensus and then comparing. You could (and most likely should) just not convert and compare the strong types instead. In fact, if there's any reason not to something is probably horribly wrong.

Reply:
I don't think we have comparison opcodes on `CompactTarget`. If we do we are planning to get rid of them. See https://github.com/rust-bitcoin/rust-bitcoin/issues/2110. (I see that you commented there that we should have explict named methods for the different comparison methods. Ok, but they don't exist right now.)

Reply:
FWIW, the pow stuff is on the top of my list at the moment, I started on it yesterday.

Reply:
Oh, crap, so we have an actual use case for comparison, kinda undermining the point of #2110, but it's still obscure. It's also super weird that Core compares those. Makes me quite intrigued to figure out WTF is going on.

Using `Eq` now could be a nice way to not forget to change the code (presumably to use explicit method) when #2110 is resolved.

Reply:
Well, yes, we have a use case *within PoW computation*. But there's lots of stuff we do in that computation that we don't necessarily want to expose by itself.